### PR TITLE
feat: new demographics sub-race questions

### PR DESCRIPTION
**Pull request by emilyjablonski**
# Pull Request Template

## Issue Overview

This PR addresses #1739

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

Changes demographics to be a series of checkboxes (with sub-categories and occasionally also a custom text input) instead of a dropdown, switches the order of ethnicity and race, and removes gender and sexual orientation from the UI.
Open to changes how the data is being stored, but I just made the race field a string array.
I had to modify the FieldGroup component in some funky ways to be able to support this. In the future we should look into not having every field in a FieldGroup have the same name for every input because that made it hard to target individual inputs to conditionally show subsections and to set default values.

## How Can This Be Tested/Reviewed?

Submit an application from the public side with a set of the new checkboxes. Ensure they exist in partners on the app page, and that you can edit them there. From partners I locally edited the CSV api call in `bloom/sites/partners/pages/listings/[id]/applications/index.tsx` on 136 and added `includeDemographics: true` to test the CSV.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx [hunk 1]
```diff
@@ -50,16 +50,20 @@ const FormDemographics = () => {
...
+              id: rootKey,
+              label: t(`application.review.demographics.raceOptions.${rootKey}`),
+              value: rootKey,
+              subFields: raceKeys[rootKey].map((subKey) => ({
+                id: subKey,
+                label: t(`application.review.demographics.raceOptions.${subKey}`),
+                value: subKey,
+              })),
+            }))}
+            type="checkbox"
```
Comment (seanmalbert) [inline sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx:65]: Should this and the FieldGroup in `review/demographics` be moved to a shared helper?

**Review by seanmalbert — CHANGES_REQUESTED** [review]
@emilyjablonski ,
This looks great, just some minor things to resolve first. Can you also check with Kathy to make sure she's ok with the format in the CSV. I originally commented in the issue with a format of `Asian[Chinese, Filipino];Black or African American;` as an example, so I want to make sure that's not what's expected.

This is also another one where we want to have tests for this, but the pieces aren't quite ready. It'll take a bit to review #2118 and I'm assuming this will get in first, so that'll have to be updated to accommodate these updates. I'm not sure if those tests are asserting what's in localStorage to get the demographic data, so we'll want to make sure we're covering that.

**Comment by emilyjablonski** [comment]
TODO for me: swap order of race and ethnicity

File: sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx (hunk 1 above)
Comment (emilyjablonski) [reply sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx:65]: At the moment we're not able to import anything from ui-components in shared-helpers due to some sort of build issue. So I added this commented out in helpers with the PR number that should fix it to return to.

**Comment by emilyjablonski** [comment]
@seanmalbert Checked in w Kathy and updated the keys to come through as root-sub (i.e. asian-filipino)

**Comment by emilyjablonski** [comment]
Updated the CSV to format the keys, so some examples of how they'll come through:

American Indian / Alaskan Native, Asian, Native Hawaiian / Other Pacific Islander, White, Other / Multiracial: Custom value, Asian[Korean], Asian[Vietnamese], Asian[Other Asian: Custom value], Native Hawaiian / Other Pacific Islander[Other Pacific Islander: Custom value]

Asian, Black / African American, Asian[Other Asian: Custom value]

Asian, White

File: shared-helpers/src/formKeys.ts [hunk 2]
```diff
@@ -127,6 +141,51 @@ export const sexualOrientation = [
...
+      if (formObject[elem].isArray) {
+        formObject[elem].forEach(() => {
+          modifiedArray.push(getValue(elem))
+        })
+      } else {
+        modifiedArray.push(getValue(elem))
+      }
+    })
+  return modifiedArray
+}
```
Comment (seanmalbert) [inline shared-helpers/src/formKeys.ts:174]: Can you write unit tests for this? It looks like the demographics use case doesn't ever hit the check for `formObject[elem].IsArray`. Do we need to support that now?

File: sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx [hunk 3]
```diff
@@ -30,8 +29,50 @@ const FormDemographics = () => {
...
+      subFields: raceKeys[rootKey].map((subKey) => ({
+        id: subKey,
+        label: t(`application.review.demographics.raceOptions.${subKey}`),
+        value: subKey,
+        defaultChecked: isKeyIncluded(subKey),
+        additionalText: subKey.indexOf("other") >= 0,
+        defaultText: getCustomValue(subKey),
+      })),
+    }))
+  }, [register])
```
Comment (seanmalbert) [inline sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx:66]: I don't think this properly handles the `Other / Multiracial` option. raceKeys doesn't have the dynamic value of `otherMultiracial`, so it never hits subFields or the `getCustomValue` call for defaultText to handle a value like `otherMultiracial: test other race`.

File: ui-components/src/forms/FieldGroup.tsx [hunk 4]
```diff
@@ -1,14 +1,20 @@
-import React from "react"
+import React, { useState, useEffect } from "react"
```
Comment (seanmalbert) [inline ui-components/src/forms/FieldGroup.tsx:1]: I think this component is a great candidate for unit testing components. With these updates, this is now probably one of the more complex field components we have, so it'd be great to get coverage on it with all its iterations.

**Review by seanmalbert — CHANGES_REQUESTED** [review]

File: shared-helpers/src/formKeys.ts (hunk 2 above)
Comment (emilyjablonski) [reply shared-helpers/src/formKeys.ts:174]: We don't support unit testing of JS in shared-helpers yet on dev, I can pull in my other branch

File: sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx (hunk 3 above)
Comment (emilyjablonski) [reply sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx:66]: Are you seeing that while testing in the UI? It seems to be functional

File: ui-components/src/forms/FieldGroup.tsx (hunk 4 above)
Comment (emilyjablonski) [reply ui-components/src/forms/FieldGroup.tsx:1]: Added more tests

File: sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx (hunk 3 above)
Comment (emilyjablonski) [reply sites/partners/src/applications/PaperApplicationForm/sections/FormDemographics.tsx]: Touched base offline, fixed!

File: shared-helpers/src/formKeys.ts (hunk 2 above)
Comment (emilyjablonski) [reply shared-helpers/src/formKeys.ts:174]: Pulled in changes and wrote tests

File: shared-helpers/__tests__/formKeys.test.ts [hunk 5]
```diff
@@ -1,6 +1,44 @@
-import * as formKeys from "../src/formKeys"
+import { cleanup } from "@testing-library/react"
```
Comment (seanmalbert) [inline shared-helpers/__tests__/formKeys.test.ts:1]: The tests look good! Can you also add tests for if `formObject[elem].isArray`?

**Review by seanmalbert — APPROVED** [review]
Hey @emilyjablonski , the updates look good, thank you! Before merging can you update the formKeys tests to cover the use-case I called out in the comments. There are also a number of `accesslint` bot comments. We should probably address some of those.

**Comment by emilyjablonski** [comment]
@seanmalbert Sounds good added another test. And for a11y issues I don't know why they're happening - that component has an aria label and a name and a label tag that is associated with the input, which I believe is what we need, and it doesn't provide other context :/

**Pull request merged**